source/extensions/transport_sockets/s2a/config.cc

```diff
@@ -0,0 +1,119 @@
+#include "source/extensions/transport_sockets/s2a/config.h"
+
+#include "envoy/extensions/transport_sockets/s2a/v3alpha/s2a.pb.h"
+#include "envoy/extensions/transport_sockets/s2a/v3alpha/s2a.pb.validate.h"
+#include "envoy/registry/registry.h"
+#include "envoy/server/transport_socket_config.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/grpc/google_grpc_context.h"
+#include "source/common/protobuf/protobuf.h"
+#include "source/common/protobuf/utility.h"
+#include "source/extensions/transport_sockets/s2a/grpc_tsi.h"
+#include "source/extensions/transport_sockets/s2a/tsi_socket.h"
+
+#include "absl/container/node_hash_set.h"
+#include "absl/strings/str_join.h"
+#include "absl/status/statusor.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace S2A {
+namespace {
+
+// Manage S2A singleton state via SingletonManager
+class S2ASharedState : public Singleton::Instance {
```

Comment:
nit: can this class be marked `final`? I'm worried about a future where we have to expose this outside to the header file and then someone overrides the destructor and we end up with a memory leak that might be hard to catch.